src/Client/HedgedConnectionsFactory.h

```diff
@@ -0,0 +1,139 @@
+#pragma once
+
+#if defined(OS_LINUX)
+
+#include <Common/TimerDescriptor.h>
+#include <Common/Epoll.h>
+#include <Common/FiberStack.h>
+#include <Common/Fiber.h>
+#include <Client/ConnectionEstablisher.h>
+#include <Client/ConnectionPoolWithFailover.h>
+#include <Core/Settings.h>
+#include <unordered_map>
+#include <memory>
+
+namespace DB
+{
+
+/** Class for establishing hedged connections with replicas.
+  * The process of establishing connection is divided on stages, on each stage if
+  * replica doesn't respond for a long time, we start establishing connection with
+  * the next replica, without cancelling working with previous one.
+  * It works with multiple replicas simultaneously without blocking by using epoll.
+  */
+class HedgedConnectionsFactory
+{
+public:
+    using ShuffledPool = ConnectionPoolWithFailover::Base::ShuffledPool;
+    using TryResult = PoolWithFailoverBase<IConnectionPool>::TryResult;
+
+    enum class State
+    {
+        READY,
+        NOT_READY,
+        CANNOT_CHOOSE,
+    };
+
+    struct ReplicaStatus
+    {
+        explicit ReplicaStatus(ConnectionEstablisherAsync connection_stablisher_) : connection_establisher(std::move(connection_stablisher_))
+        {
+        }
+
+        ConnectionEstablisherAsync connection_establisher;
+        TimerDescriptor change_replica_timeout;
+        bool is_ready = false;
+    };
+
+    HedgedConnectionsFactory(const ConnectionPoolWithFailoverPtr & pool_,
+                        const Settings * settings_,
+                        const ConnectionTimeouts & timeouts_,
+                        std::shared_ptr<QualifiedTableName> table_to_check_ = nullptr);
+
+    /// Create and return active connections according to pool_mode.
+    std::vector<Connection *> getManyConnections(PoolMode pool_mode);
+
+    /// Try to get connection to the new replica. If start_new_connection is true, we start establishing connection
+    /// with the new replica. Process all current events in epoll (connections, timeouts),
+    /// if there is no events in epoll and blocking is false, return NOT_READY.
+    /// Returned state might be READY, NOT_READY and CANNOT_CHOOSE.
+    /// If state is READY, replica connection will be written in connection_out.
+    State waitForReadyConnections(bool blocking, Connection *& connection_out);
```

Comment:
`blocking` flag is still not needed externally. It is called only once from `HedgedConnections` with `blocking = false`.
We can add `waitForReadyConnectionsImpl` with flag and remove it from `waitForReadyConnections`.
(it is not very important, tough)